Github API reverse proxy HTTP cache

cjwagner opened the pull request:
Previous PR using nginx: #7603 
Neither nginx nor varnish respect the [HTTP caching RFC](https://tools.ietf.org/html/rfc7234) with respect to forcing revalidation of cache entries using `max-age=0`, `no-cache`, or `must-revalidate` (this is surprising, these are designed to be HTTP caches!) Neither of these caches support caching a response if it must be revalidated in the future.

Since this is the primary requirement of a Github API cache they aren't a valid option so I pieced together an HTTP reverse proxy cache in go. It enforces the caching policy we require to ensure revalidation and also includes prometheus instrumentation and concurrent request coalescing.

/cc @stevekuznetsov @BenTheElder 
/area prow

BenTheElder commented on ghproxy/.gitignore:
```diff
@@ -0,0 +1 @@
+ghproxy
```
nit: newline? /shrug

cjwagner replied:
Done.

BenTheElder commented on ghproxy/ghcache/coalesce.go:
```diff
@@ -0,0 +1,95 @@
+	"sync"
+)
+
+// requestCoalescer allows concurrent requests for the same key to share a
```
how often do we expect this to occur?

BenTheElder replied:
what is the RTT like? this still feels like something the cache should handle, without needing a global lock or any of the other complexity. this feels like purely premature optimization. I foresee hung goroutines.

cjwagner replied:
RTT is between 0.25s and 0.60s from the Prow cluster.
I agree that this is something the cache should handle. The vendored caching layer doesn't handle it itself so I wrapped it with a transport so that the ghcache handles it transparently.

I don't think this is a premature optimization. This is a common case for the most used API calls like `GetIssueLabels` and `ListIssueComments`. The logs show that there are frequently multiple calls for the same resource occurring within a second (the time granularity that stackdriver displays). Here is a sample showing just the `GetIssueLabels` calls for an event:
```
2018-04-24 17:12:15.000 PDT  {"component":"hook","msg":"GetIssueLabels(kubernetes, kubernetes, 62297)","client":"github","level":"info"} 
2018-04-24 17:12:15.000 PDT  {"component":"hook","msg":"GetIssueLabels(kubernetes, kubernetes, 62297)","client":"github","level":"info"} 
2018-04-24 17:12:15.000 PDT  {"component":"hook","msg":"GetIssueLabels(kubernetes, kubernetes, 62297)","client":"github","level":"info"} 
2018-04-24 17:12:15.000 PDT  {"component":"hook","msg":"GetIssueLabels(kubernetes, kubernetes, 62297)","client":"github","level":"info"} 
2018-04-24 17:12:15.000 PDT  {"msg":"GetIssueLabels(kubernetes, kubernetes, 62297)","client":"github","level":"info","component":"hook"} 
2018-04-24 17:12:15.000 PDT  {"client":"github","level":"info","component":"hook","msg":"GetIssueLabels(kubernetes, kubernetes, 62297)"} 
2018-04-24 17:12:15.000 PDT  {"component":"hook","msg":"GetIssueLabels(kubernetes, kubernetes, 62297)","client":"github","level":"info"} 
2018-04-24 17:12:15.000 PDT  {"component":"hook","msg":"GetIssueLabels(kubernetes, kubernetes, 62297)","client":"github","level":"info"} 
```
Without the synchronization layer I added, all of the above calls would use an API token if the cache is stale instead of just using one token in total.

I'm not too worried about the global lock. We don't hold it while executing requests, only to lookup if a request is in flight and add/delete entries. So requests for different paths won't block each other for more than the time it takes to update the map (which is negligible when compared to RTT time to Github).
I can change the map to be a concurrent `sync.Map` if that is preferable though.

BenTheElder replied:
> I'm not too worried about the global lock. We don't hold it while executing requests, only to lookup if a request is in flight and add/delete entries. So requests for different paths won't block each other for more than the time it takes to update the map (which is negligible when compared to RTT time to Github).

I'm much more concerned with all the waiters etc. That code looks ripe for bugs / hanging.

> I can change the map to be a concurrent sync.Map if that is preferable though.

sync.Map is not suited to this. it's a rather complex map internally only optimized for specific use cases (maps that are append-only and read heavy, or also maps that are only a ~ fixed set of keys where only the values are touched, for some patterns)

BenTheElder replied:
similarly https://github.com/kubernetes/test-infra/pull/7848/files#r183921856 it looks to me like the keys map in requestCoalescer is append-only right now but probably shouldn't be..

cjwagner replied:
I added a unit test. Let me know what you think.

cjwagner commented on ghproxy/ghproxy.go:
```diff
@@ -0,0 +1,159 @@
+func flagOptions() *options {
+	o := &options{}
+	flag.StringVar(&o.dir, "cache-dir", "", "Directory to cache to if using a disk cache.")
+	flag.IntVar(&o.sizeM, "cache-sizeM", 0, "Cache size in Mb if using a disk cache.")
```
Switched everything to GB.

cblecker commented:
So for the edification of someone watching this from a distance..
If I understand the problem correctly, if Github responds with a `max-age=0`, `no-cache`, or `must-revalidate` varnish and nginx will decide that the response is ineligible for caching at all and won't cache it (even with the intention of doing a future conditional request). Because of this, you've created your own proxy.

Is that accurate? If so, do you know what kind of requests generate a `max-age=0`? Most of the ones I've ever seen have a max-age of 60.

cjwagner commented:
@cblecker Yes, that is mostly accurate. Github **is** sending `max-age=60` for all the requests that I have seen. That isn't acceptable for us though. Prow likely won't tolerate 60 seconds of staleness so I want to force revalidation of cache entries by overwriting the cache-control policy.

cblecker commented:
@cjwagner Instead of setting to zero, would 1s of staleness be tolerable? Even if it's within that 60s, shouldn't it revalidate with conditional requests anyways?

BenTheElder commented on ghproxy/ghcache/coalesce.go:
```diff
@@ -0,0 +1,95 @@
+// single upstream request and response.
+type requestCoalescer struct {
+	sync.Mutex
+	keys map[string]*responseWaiter
```
I don't see anything ever GCing entries from this map.

cjwagner replied:
We GC on line 85 when the request returns. Entries only exist while the request is in flight.

cjwagner replied:
We do now. I added a timeout handler to the reverse proxy.

BenTheElder commented on ghproxy/ghproxy.go:
```diff
@@ -0,0 +1,159 @@
+		if err != nil {
+			logger.WithError(err).Error("Failed to get disk metrics")
+		} else {
+			diskFree.Set(float64(bytesFree) / 1e9)
```
this is not gibibytes. gibi is 2^30. why not use gigabytes? storage is generally measured / sold in GB.

stevekuznetsov replied:
Wow, I had that backwards ... I could have sworn mega/giga were the powers of 2 and mebi/gibi were powers of ten -.-

BenTheElder commented on ghproxy/ghcache/ghcache.go:
```diff
@@ -0,0 +1,156 @@
+		diskv.New(diskv.Options{
+			BasePath:     path.Join(cacheDir, "data"),
+			TempDir:      path.Join(cacheDir, "temp"),
+			CacheSizeMax: uint64(cacheSizeGB) * uint64(1000000000), // convert G to B
```
This is correct as-is.

BenTheElder commented on prow/cluster/gce-ssd-retain_storageclass.yaml:
```diff
@@ -0,0 +1,24 @@
+provisioner: kubernetes.io/gce-pd
+parameters:
+  type: pd-ssd
+reclaimPolicy: Retain
```
newline

cjwagner replied:
Done.

cjwagner commented:
> Instead of setting to zero, would 1s of staleness be tolerable?

I considered that alternative (I was able to get both nginx and varnish to do that), but I'm not sure if it is tolerable and if its not it would introduce some nasty bugs. As shown in the log samples above, Prow does a lot within the first few seconds an event is received so 1s of staleness is worrying.
> Even if it's within that 60s, shouldn't it revalidate with conditional requests anyways?

According to the HTTP caching RFC, not unless `must-revalidate` is specified in a cached response header or `max-age=0` is specified in a request header. Neither nginx nor vanish respect these though.

cjwagner commented on ghproxy/ghcache/coalesce.go:
```diff
@@ -0,0 +1,95 @@
+//  or
+// 2) A function that should be used to return a response to other requesters
+//    after the caller completes the request themselves.
+func (r *requestCoalescer) Request(key string) (*http.Response, func(*http.Response, error), error) {
```
Done.

cjwagner commented on ghproxy/ghcache/coalesce.go:
```diff
@@ -0,0 +1,95 @@
+	err  error
+}
+
+// Request returns one of two things depending on if it there is already a
```
I changed the requestCoalescer to implement the RoundTripper interface instead of trying to separate the synchronization from the request handling. It has a consistent interface and should be clearer now.

stevekuznetsov commented on ghproxy/ghcache/coalesce.go:
```diff
@@ -0,0 +1,108 @@
+	waiter.L.Unlock()
+
+	cacheCounter.WithLabelValues(cacheResponseMode(resp.Header)).Inc()
+	return resp, err
```
Not sure if we care but in this case you get the response regardless of whether `err == nil` whereas if you get coalesced you get _either_ the response or the error.

cjwagner replied:
Fixed.

stevekuznetsov commented on ghproxy/ghcache/coalesce.go:
```diff
@@ -0,0 +1,108 @@
+		waiter.L.Lock()
+		r.Unlock()
+		waiter.waiting = true
+		waiter.Wait()
```
Do we need to worry about this from GoDoc?

> Because c.L is not locked when Wait first resumes, the caller typically cannot assume that the condition is true when Wait returns. Instead, the caller should Wait in a loop.

cjwagner replied:
No it doesn't apply to this use case because the condition we are waiting for remains true once it becomes true. 
I added a comment.

stevekuznetsov commented:
/hold

cjwagner commented:
/retest

cjwagner commented:
/test all

cjwagner commented:
Comments addressed, PTAL.

I also made #8050 so that we can configure Prow components to bypass the cache if it is unavailable for any reason.

cjwagner commented:
Assigned the cache to a dedicated node and added fall back endpoints to components configured to use the cache.
PTAL

BenTheElder commented:
skimmed it, will do another pass later ... or monday? I would consider adding a readme documenting how / why this is setup a bit.

cjwagner commented:
Added some simple READMEs.

cjwagner commented:
Rebased, improved logging, and fixed a bug (only try to close the request body if it is non-nil).
This is now deployed and appears to be working well: http://velodrome.k8s.io/dashboard/db/github-cache?refresh=1m&orgId=1

PTAL
/hold cancel

BenTheElder commented:
/assign
will look at this today

BenTheElder commented:
http://velodrome.k8s.io/dashboard/db/github-cache?refresh=1m&orgId=1

cjwagner commented:
/hold cancel
